File: C++/jumpSearch.cpp
// Jump search algorithm
// This algorithm finds the index of an element in a sorted array
// Time complexity : between Linear Search ( ( O(n) ) and Binary Search ( O (Log n) )

#include<iostream>
#include<math.h>

using namespace std;

int jumpSearch(int arr[], int length, int ele)
{
	int left, right;
	left = 0;
	right = sqrt(length);
	
	while(arr[right] <= ele && right < length)
	{
		left = right;
		right +=sqrt(length);
	}
	
	for(int i=left; i<right; i++)
	{
		if(arr[i] == ele)
		{
			return i+1;
			break;
		}
	}
	
	return -1;
}

int main()
{
	int ar[10],i,n;
	int ele;
	
	cout<<"Number of elements: ";
	cin>>n;
	cout<<"Enter array elements: ";
	for(i=0;i<n;i++)
	{
		cin>>ar[i];
	}
	
	cout<<"Entered array elements: ";
	for(i=0;i<n;i++)
	{
		cout<<ar[i]<<" ";
	}
	cout<<endl;
	cout<<"Enter the element to be search: ";
	cin>>ele;
	
	int result = jumpSearch(ar, n, ele);
	
	cout<<"Result is: "<<result;
	
	return 0;
}

File: JavaScript/jumpSearch.js
function jumpSearch(arr, x, n) {
    // Finding block size to be jumped
    let step = Math.sqrt(n);

    // Finding the block where element is present
    let prev = 0;
    while (arr[Math.min(step, n) - 1] < x) {
        prev = step;
        step += Math.sqrt(n);
        if (prev >= n) {
            return -1;
        } 
    }

    // Doing a linear search for x in 
    // block beginning with prev

    while (arr[prev] < x) {
        prev += 1;
        if (prev == Math.min(step, n)) {
            return -1;
        }
    }

    if (arr[prev] == x) {
        return prev;
    }
    return -1;
}

let arr = [ 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610 ];
let x = 55;
let n = arr.length;

let index = jumpSearch(arr, x, n);
console.log(`Number ${x} is at index ${index}`);